Article Update: Profiling Web Audio apps in Chrome (#8101)
GoogleChrome/web.dev: src/site/content/en/blog/profiling-web-audio-apps-in-chrome/index.md

* **Time slices spent by specific function calls** on different threads
* **Audio callback timing** in the timeline view

It usually shows missed audio callback deadlines or big garbage collection that
might cause unexpected audio glitches. This information is useful for

> understanding a underlying problem, so Chromium engineers will often ask it

Reviewer: ```suggestion
understanding an underlying problem, so Chromium engineers will often ask it
```

especially when the local reproduction is not feasible. The general

> instructions for tracing is available [here](chromium-tracing).

Reviewer: ```suggestion
instructions for tracing are available [here](chromium-tracing).
```

### When do you use Web Audio DevTools extension?

When you want to visualize the audio graph and monitor how the audio renderer

> performs in real time. The audio graph, a network of AudioNode objects to

Reviewer: ```suggestion
performs in real time. The audio graph, a network of `AudioNode` objects to
```

generate and synthesize an audio stream, often gets complex but the graph

> topology is opaque by design. (Web Audio API doesn’t have features for

Reviewer: ```suggestion
topology is opaque by design. (The Web Audio API doesn’t have features for
```

{% Img src="image/tcFciHGuF3MxnTr1y5ue01OGLBn2/1UN5udXOW56ooihw5M18.png", alt="Screen shot comparing unstable vs stable callback timing.", width="800", height="252" %}

**Your options:**

* Increase the system audio callback buffer size by adjusting the
  [`latencyHint`](https://webaudio.github.io/web-audio-api/#dom-audiocontextoptions-latencyhint)
  option.
* If you find this problem, [file an issue on crbug.com](https://bugs.chromium.org/p/chromium/issues/entry?components=Blink%3EWebAudio) 
  with the tracing data.

## Use Web Audio DevTools Extension

You also can use the DevTools extension specifically designed for Web Audio API.
Unlike the tracing tool, this provides real time inspection of graphs and
performance metrics.

This extension needs to be installed from
[Chrome Web Store](https://chrome.google.com/webstore/detail/audion/cmhomipkklckpomafalojobppmmidlgl).

After the installation, you access the panel by opening Chrome DevTools and
clicking “Web Audio” on the top menu.

> {% Img src="", alt="", width="800", height="" %}

Reviewer 2: ```suggestion
{% Img src="image/vvhSqZboQoZZN9wBvoXq72wzGAf1/JUXLwtX83TOXqRHD2CGU.png", alt="Screenshot showing how to open Web Audio panel in Chrome DevTools.", width="800", height="201" %}

Author: Thanks! Done.

The Web Audio panel has 4 components: context selector, property inspector,
graph visualizer, and performance monitor.

> ### Context Selector

Reviewer 2: ```suggestion
{% Img src="image/vvhSqZboQoZZN9wBvoXq72wzGAf1/4xQhLJBdDWW3X2nIssIS.png", alt="Screenshot of the Web Audio panel in Chrome DevTools", width="800", height="416" %}

### Context Selector
```

Author: Done.

> Since a page can have multiple BaseAudioContexts, this drop-down menu allows you

Reviewer 2: ```suggestion
Since a page can have multiple `BaseAudioContexts`, this drop-down menu allows you

Author: Done.

  quality. (See the example 3 above)

* **Render Capacity** (percent): When the capacity gets close to 100 percent, it
  means that the renderer is doing too much for a given render budget, so you
  should consider doing less (e.g. using fewer `AudioNodes` in the
  graph).

You can manually trigger a garbage collector by clicking the trash can icon.

## Legacy WebAudio DevTools panel

The extension is now a recommended method by the Chrome Web Audio team, but the
legacy WebAudio DevTools panel is also available. You can access this panel by
clicking the "three dots" menu in the top right corner of DevTools, then go to
**More tools** then **WebAudio**.

{% Img src="image/tcFciHGuF3MxnTr1y5ue01OGLBn2/t2eX431PTio5oOFkmOtR.jpg", alt="Screen shot showing how to open WebAudio panel in Chrome DevTools.", width="800", height="423" %}

## Conclusion

Debugging audio is hard. Debugging audio in the browser is even harder. However,
these tools can ease the pain by providing you with useful insights on how the
web audio code performs. In some cases, however, you may find problems in Chrome
or the extension. Then do not be afraid to

> [file a bug on crbug.com][r-file-a-bug] or on the

Reviewer 2: ```suggestion
[file a bug on crbug.com][cr-file-a-bug] or on the

Author: Done.

{% Img src="image/tcFciHGuF3MxnTr1y5ue01OGLBn2/1UN5udXOW56ooihw5M18.png", alt="Screen shot comparing unstable vs stable callback timing.", width="800", height="252" %}

**Your options:**

* Increase the system audio callback buffer size by adjusting the
  [`latencyHint`](https://webaudio.github.io/web-audio-api/#dom-audiocontextoptions-latencyhint)
  option.

> * If you find this problem, [file an issue on crbug.com](https://bugs.chromium.org/p/chromium/issues/entry?components=Blink%3EWebAudio) 

Reviewer: ```suggestion
* If you find a problem, [file an issue on crbug.com](https://bugs.chromium.org/p/chromium/issues/entry?components=Blink%3EWebAudio) 
```

  with the tracing data.

> ## Use Web Audio DevTools Extension

Reviewer: ```suggestion
## Use the Web Audio DevTools extension
```

> You also can use the DevTools extension specifically designed for Web Audio API.

Reviewer: ```suggestion
You can also use the DevTools extension specifically designed for the Web Audio API.
```

Unlike the tracing tool, this provides real time inspection of graphs and
performance metrics.

> This extension needs to be installed from

Reviewer: ```suggestion
This extension needs to be installed from the
```

[Chrome Web Store](https://chrome.google.com/webstore/detail/audion/cmhomipkklckpomafalojobppmmidlgl).

After the installation, you access the panel by opening Chrome DevTools and
clicking “Web Audio” on the top menu.

{% Img src="image/vvhSqZboQoZZN9wBvoXq72wzGAf1/JUXLwtX83TOXqRHD2CGU.png", alt="Screenshot showing how to open Web Audio panel in Chrome DevTools.", width="800", height="201" %}

The Web Audio panel has 4 components: context selector, property inspector,
graph visualizer, and performance monitor.

> {% Img src="image/vvhSqZboQoZZN9wBvoXq72wzGAf1/4xQhLJBdDWW3X2nIssIS.png", alt="Screenshot of the Web Audio panel in Chrome DevTools", width="800", height="416" %}

Reviewer: ```suggestion
{% Img src="image/vvhSqZboQoZZN9wBvoXq72wzGAf1/4xQhLJBdDWW3X2nIssIS.png", alt="Screenshot of the Web Audio panel in Chrome DevTools.", width="800", height="416" %}
```

### Context Selector

> Since a page can have multiple `BaseAudioContext`s, this drop-down menu allows you

Reviewer: ```suggestion
Since a page can have multiple `BaseAudioContext` objects, this drop-down menu allows you
```

> to choose the context you want to inspect. You can also manually trigger garbage

Reviewer: ```suggestion
choosing the context you want to inspect. You can also manually trigger garbage
```

collection by clicking the trash can icon on the left. 

### Property Inspector

> This side panel shows various properties of a user-selected context or

Reviewer: ```suggestion
The side panel shows various properties of a user-selected context or
```

`AudioNode`. Inspecting dynamic values in `AudioParam` is not supported.

### Graph Visualizer

This view renders the current graph topology of a user-selected context. This
visualization changes dynamically in real time. By clicking an element in the
visualization, you can inspect the detailed information on the property
inspector.

### Performance Monitor

> This status bar at the bottom is only active when the selected BaseAudioContext

Reviewer: ```suggestion
The status bar at the bottom is only active when the selected `BaseAudioContext`
```

> is an AudioContext, which runs in real-time. This bar shows the instantaneous

Reviewer: ```suggestion
is an `AudioContext`, which runs in real-time. This bar shows the instantaneous
```

> audio stream quality of a selected AudioContext and is updated every second. It

Reviewer: ```suggestion
audio stream quality of a selected `AudioContext` and is updated every second. It
```

provides the following information: 

* **Callback interval** (ms): Displays the weighted mean/variance of callback
  interval. Ideally the mean should be stable and the variance should be close
  to zero. If you see a large variance, it means that the system-level audio
  callback function has unstable timing that can lead to poor audio stream

>   quality. (See the example 3 above)

Reviewer: ```suggestion
  quality. (See example 3, above.)
```

* **Render Capacity** (percent): When the capacity gets close to 100 percent, it
  means that the renderer is doing too much for a given render budget, so you

>   should consider doing less (e.g. using fewer `AudioNodes` in the

Reviewer: ```suggestion
  should consider doing less (e.g. using fewer `AudioNode` objects in the
```

  graph).

You can manually trigger a garbage collector by clicking the trash can icon.

## Legacy WebAudio DevTools panel

The extension is now a recommended method by the Chrome Web Audio team, but the
legacy WebAudio DevTools panel is also available. You can access this panel by

> clicking the "three dots" menu in the top right corner of DevTools, then go to

Reviewer: ```suggestion
clicking the "three dots" menu in the top right corner of DevTools, then going to
```

> **More tools** then **WebAudio**.

Reviewer: ```suggestion
**More tools**, then **WebAudio**.
```
